otherlibs/systhreads/st_win32.h

```diff
@@ -24,7 +24,242 @@ Caml_inline void st_msleep(int msec)
   Sleep(msec);
 }
 
-#include "st_pthreads.h"
+/* POSIX thread implementation of the "st" interface */
+
+#include <errno.h>
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <signal.h>
+#include <time.h>
+#include <caml/osdeps.h>
+
+
+typedef HANDLE st_thread_id;
+
+/* Thread creation. Created in detached mode if [res] is NULL. */
+static int st_thread_create(st_thread_id * res,
+                            unsigned ( WINAPI *start_address )( void * ),
+                            void * arg)
+{
+  st_thread_id thr;
+  thr = (st_thread_id) _beginthreadex(
+    NULL, /* security: handle can't be inherited */
+    0,    /* stack size */
+    start_address,
+    arg,
+    0,    /* run immediately */
+    NULL  /* thread identifier */
+    );
+  if (thr == 0) {
+    return errno;
+  } else if (res == NULL) {
+    /* detach */
+    if (!CloseHandle(thr)) {
+      int err = caml_posixerr_of_win32err(GetLastError());
+      return err == 0 ? EINVAL : err;
+    }
+  } else {
+    *res = thr;
+  }
+  return 0;
+}
+
+/* Thread termination */
+
+static void st_thread_join(st_thread_id thr)
+{
+  WaitForSingleObject(thr, INFINITE);
+  /* best effort: ignore errors */
+}
+
+/* Thread-specific state */
+
+typedef DWORD st_tlskey;
+
+static int st_tls_newkey(st_tlskey * res)
+{
+  DWORD index = TlsAlloc();
+  if (index == TLS_OUT_OF_INDEXES) {
+    int err = caml_posixerr_of_win32err(GetLastError());
+    return err == 0 ? EINVAL : err;
+  }
+  *res = index;
+  return 0;
+}
+
+Caml_inline void * st_tls_get(st_tlskey k)
+{
+  /* No errors are returned from pthread_getspecific(). Ignore
+   * TlsGetValue() errors.  */
+  return TlsGetValue(k);
+}
+
+Caml_inline void st_tls_set(st_tlskey k, void * v)
+{
+  (void)TlsSetValue(k, v);
+}
+
+/* The master lock.  This is a mutex that is held most of the time,
+   so we implement it in a slightly convoluted way to avoid
+   all risks of busy-waiting.  Also, we count the number of waiting
+   threads. */
+
+typedef struct {
+  bool init;                      /* have the mutex and the cond been
+                                     initialized already? */
+  SRWLOCK lock;                   /* to protect contents */
+  bool busy;                      /* false = free, true = taken */
+  atomic_uintnat waiters;         /* number of threads waiting on master lock */
+  CONDITION_VARIABLE is_free;     /* signaled when free */
+} st_masterlock;
+
+/* Returns non-zero on failure */
+static int st_masterlock_init(st_masterlock * m)
+{
+  if (!m->init) {
+    InitializeSRWLock(&m->lock);
+    InitializeConditionVariable(&m->is_free);
+    m->init = true;
+  }
+  m->busy = true;
+  atomic_store_release(&m->waiters, 0);
+  return 0;
+}
+
+static uintnat st_masterlock_waiters(st_masterlock * m)
+{
+  return atomic_load_acquire(&m->waiters);
+}
+
+static void st_masterlock_acquire(st_masterlock *m)
+{
+  AcquireSRWLockExclusive(&m->lock);
+  while (m->busy) {
```

Comment:
I am worried an overzealous compiler (i.e. `clang-cl`) could optimize this into `if (m->busy) while (1)`, because `busy` is marked neither `volatile` nor `atomic`.

I know this logic was already present in the existing code, but I'm worried nevertheless.

Reply:
I've marked both boolean fields as atomic, and also in the pthreads code.